compiler/noirc_frontend/src/resolve_locations.rs

```diff
@@ -53,28 +57,49 @@ impl NodeInterner {
     /// For a given [Index] we return [Location] to which we resolved to
     /// We currently return None for features not yet implemented
     /// TODO(#3659): LSP goto def should error when Ident at Location could not resolve
-    fn resolve_location(&self, index: impl Into<Index>) -> Option<Location> {
+    fn resolve_location(
+        &self,
+        index: impl Into<Index>,
+        return_type_location_instead: bool,
+    ) -> Option<Location> {
         let node = self.nodes.get(index.into())?;
 
         match node {
-            Node::Function(func) => self.resolve_location(func.as_expr()),
-            Node::Expression(expression) => self.resolve_expression_location(expression),
+            Node::Function(func) => {
+                self.resolve_location(func.as_expr(), return_type_location_instead)
+            }
+            Node::Expression(expression) => {
+                self.resolve_expression_location(expression, return_type_location_instead)
+            }
             _ => None,
         }
     }
 
     /// Resolves the [Location] of the definition for a given [HirExpression]
     ///
     /// Note: current the code returns None because some expressions are not yet implemented.
-    fn resolve_expression_location(&self, expression: &HirExpression) -> Option<Location> {
+    fn resolve_expression_location(
+        &self,
+        expression: &HirExpression,
+        return_type_location_instead: bool,
+    ) -> Option<Location> {
         match expression {
             HirExpression::Ident(ident) => {
                 let definition_info = self.definition(ident.id);
                 match definition_info.kind {
                     DefinitionKind::Function(func_id) => {
                         Some(self.function_meta(&func_id).location)
                     }
-                    DefinitionKind::Local(_local_id) => Some(definition_info.location),
+                    DefinitionKind::Local(local_id) => {
+                        if return_type_location_instead {
+                            match self.id_type(local_id?) {
+                                Type::Struct(struct_type, _) => Some(struct_type.borrow().location),
+                                _ => None,
+                            }
+                        } else {
+                            Some(definition_info.location)
+                        }
```

Comment:
Location handling from `resolve_expression_location` should probably be removed, it is a bit indirect and won't work for locals with no `local_id` since we get the type via `self.id_type(local_id?)`. Also, since we get the type of the definition here, it wouldn't work if the user is resolving a generic type that is known at the call site but not in the definition.

Reply:
addressed in 
14202f1: chore: extract type resolution to top